### PR TITLE
fix(pipelines): specifying the Action Role for CodeBuild steps

### LINKED ISSUE #18291: (pipelines): role on CodeBuildStep doesn't get used on CodeBuildAction
### What is the problem?

When passing a role to the CodeBuildStep the role is used on the project, but not the CodeBuildAction.

This results in a role for each CodeBuildAction being created which can cause `Maximum policy size of 10240 bytes exceeded for` on the Pipeline Default Role if you have a lot of CodeBuildSteps.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/pipelines/lib/codepipeline/_codebuild-factory.ts#L305-L322

is missing `role: this.props.role` to leverage that role.

### Reproduction Steps

This is a quick reproduction step typically the CodeBuildStep would be used as a pre or post on a stage or wave.

```
  const buildRole = new iam.Role(
    pipelineStack,
    'BuildRole',
    {
      roleName: 'BuildRole',
      assumedBy: new iam.ServicePrincipal('codebuild.amazon.com'),
    },
  );
  new cdkp.CodePipeline(pipelineStack, 'Pipeline', {
    synth: new cdkp.CodeBuildStep('Synth', {
      commands: ['/bin/true'],
      input: cdkp.CodePipelineSource.gitHub('test/test', 'main'),
      role: buildRole,
    }),
  });
```

### What did you expect to happen?

Expect to have the role used for Project and the build action.
```
  Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodeBuild::Project', {
    ServiceRole: {
      'Fn::GetAtt': [
        'BuildRole41B77417',
        'Arn',
      ],
    },
  });

  expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
    Stages: [
      // source stage
      {},
      // build stage,
      {
        Actions: [
          {
            ActionTypeId: {
              Category: 'Build',
              Owner: 'AWS',
              Provider: 'CodeBuild',
            },
            RoleArn: {
              'Fn::GetAtt': [
                'BuildRole41B77417',
                'Arn',
              ],
            },
          },
        ],
      },
    ],
  });

```

### What actually happened?

The role is used for the project but not the Code Build Action.  The Code Build Action gets a different role which is created for each build action.

for example
```
                 "RoleArn": {
                      "Fn::GetAtt": [
                        "PipelineBuildSynthCodePipelineActionRole4E7A6C97",
                        "Arn"
                      ]
                    },
```

### CDK CLI Version

N/A

### Framework Version

_No response_

### Node.js Version

N/A

### OS

N/A

### Language

Typescript

### Language Version

_No response_

### Other information

This issue can be seen directly in testing the code build step.

### PR DESCRIPTION
**Pull request by tobytipton**
This fix should address the issue #18291 

fixes #18291 
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by rix0rrr — CHANGES_REQUESTED** [review]
These should be two different roles. Please add a separate configuration field.

But this actually doesn't seem like the right fix. Can we look at the `Maximum policy size of 10240 bytes exceeded` error a little closer ?

How many roles do we have in there? Why is the policy so large?

**Comment by tobytipton** [comment]
I can change it to a `buildActionRole` property, if that makes better sense.

The `Maximum policy size of 10240 bytes exceeded` happens on the policy associated with the role which the Pipeline itself runs as because for each build role which is created the pipeline role needs to have the ability to assume that role, and because the [CodeBuildAction](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-codepipeline-actions.CodeBuildActionProps.html#role) will create a new role which means if you have to a lot of build actions to multiple stages that will cause those multiple roles to be added to the policy.

Scenario would be having a pipeline which deploys to multiple stages and each stage a code build action to do a test or change.

**Comment by saltman424** [comment]
@tobytipton if you are going to introduce a new property, you will need to either make that property default to using this.props.role, or else specify that property here (and possibly other places): https://github.com/aws/aws-cdk/blob/72941180a7188e5560a58f1509554ef038544ec4/packages/%40aws-cdk/pipelines/lib/codepipeline/codepipeline.ts#L677-L687

@rix0rrr As a note, it does seem that the intention, in at least the following case, is for them to be the same: https://github.com/aws/aws-cdk/blob/72941180a7188e5560a58f1509554ef038544ec4/packages/%40aws-cdk/pipelines/lib/codepipeline/codepipeline.ts#L807-L812

**Comment by rix0rrr** [comment]
Is this change still necessary after https://github.com/aws/aws-cdk/pull/19114 ?

**Comment by saltman424** [comment]
@rix0rrr yes, but less often. If you have enough assets, you can still get too big of a policy document from having so many roles that the pipeline role needs to be able to assume. However, with #19114 those roles are all listed in a single policy statement rather than many, so the number of assets needed to hit the limit is now a lot larger, but I believe I hit the limit when I tried something similar to #19114 with about 150 or so assets.

**Comment by ekeyser** [comment]
Thanks for checking. I'm planning on implementing the feature flag for #19114 today or tomorrow and then attempting to test this weekend. I'll update this ticket with my findings. I wanted to wait a few days.

**Comment by ekeyser** [comment]
Yes, this change is still required. I deployed a stack using 1.150.0 with the  iam minimize feature flag and still get errors regarding policy size exceeding 10240 bytes. Additionally, the `role.without_policy_updates` method appears to have in influence on what role is used for the pipeline.

File: packages/@aws-cdk/pipelines/lib/codepipeline/codebuild-step.ts [hunk 1]
```diff
@@ -66,6 +66,13 @@ export interface CodeBuildStepProps extends ShellStepProps {
    */
   readonly role?: iam.IRole;
 
+  /**
+   * Custom execution role to be used for the Code Build Action
+   *
+   * @default - A role is automatically created
+   */
+  readonly buildActionRole?: iam.IRole;
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/codepipeline/codebuild-step.ts:74]: Please rename this to `actionRole`.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I am actively working on https://github.com/aws/aws-cdk/pull/20189, which will remove the need for manual control of the action role to stay within policy sizes. 

We can still decide to merge this for environments where customers aren't allowed to create IAM roles, and they'll need a mechanism to pass in their custom roles.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**